### DISCUSSION TITLE
Profiling a native app without modifying its code

### DISCUSSION THREAD

**Discussion by pnf** [async-profiler/async-profiler#922] [Q&A] [answered]
I've been playing with doing this via LD_PRELOAD of a library with a `__attribute__((constructor))` function that starts a pthread to initialize, start sampling, then dump stacks after a while. (In fact, I just added this function to asprof.cpp itself and preload libAsyncprofiler itself.) My hack actually seems to work... except that sometimes it causes the process to core while parsing libraries.  This happens less often if the pthread waits a second before initializing, so I imagine the problem has something to do with interfering between the init hook and library loading in other threads. 
I can do more debugging, but I wanted to ask first if I'm doing something fundamentally dumb.

**Comment by apangin** [comment]
I wonder why you modify asprof? async-profiler should be LD_PRELOAD-able out of the box.
See https://github.com/async-profiler/async-profiler/issues/751#issuecomment-1587546709
If it's not - it's likely an async-profiler bug. Please try and report back.

**Comment by pnf** [comment]
It’s preloadable, but how will it start profiling and start unless someone calls asprof_init and asprof_execute?  You write in https://github.com/async-profiler/async-profiler/issues/751, that profiling must be “controlled programmatically from the process being profiled.”
 I can’t modify the executable easily, so the control must be injected somehow; static initialization was the most obvious place to do it. 

Le Jeu, 2 mai, 2024, à 19:21, Andrei Pangin a écrit:

**Reply by apangin** [reply, marked as answer]
Use `ASPROF_COMMAND` environment variable as shown in the [example](https://github.com/async-profiler/async-profiler/issues/751#issuecomment-1587546709).

**Reply by pnf** [reply]
Sorry, I didn't notice that example at the end of the discussion thread, and ended up more or less duplicating what you do in hooks.cpp for `dlopen` et al, via 2nd preloaded `.so` shim...  which I can throw away now.  Thanks.

**Reply by apangin** [reply]
Glad to know it helped. Would love to hear feedback on native app profiling experience.

**Reply by pnf** [reply]
So far so good.  I have some work to do to combine the pure-native sampling into our 
[distributed plant](https://github.com/async-profiler/async-profiler/issues/404#issuecomment-2094260048) - the hard issues always show up at scale.
